docs/source/development/AlgorithmMPISupport.rst

```diff
@@ -0,0 +1,464 @@
+
+.. code-block:: c++
+  :linenos:
+
+  #include "MantidTestHelpers/ParallelAlgorithmCreation.h"
+  #include "MantidTestHelpers/ParallelRunner.h"
+
+  namespace {
+  void run_algorithm(const Parallel::Communicator &comm,
+                     const MyType1 &arbitrary, const MyType2 &arguments) {
+    // Creating the algorithm with this helper function is the recommended way,
+    // otherwise the communicator has to be set by hand and the name of the
+    // output workspace must be set to a different value depending on comm.rank()
+    // to avoid clashes, since the threading backend in ParallelRunner shares the
+    // ADS for all 'ranks'.
+    auto alg = ParallelTestHelpers::create<Mantid::Algorithms::MyAlg>(comm);
+    alg->setProperty("InputWorkspace", boost::make_shared<WorkspaceTester>());
+    alg->execute();
+    Workspace_const_sptr ws = alg->getProperty("OutputWorkspace");
+    TS_ASSERT_EQUALS(ws->storageMode(), Parallel::StorageMode::Distributed);
+  }
+  
+  class MyAlgTest : public CxxTest::TestSuite {
+  public:
+    // ...
+  
+    void test_parallel() {
+      // Runs run_algorithm in multiple threads. The first argument passed to
+      // run_algorithm is of type Parallel::Communicator and is guaranteed to
+      // have size() > 1, i.e., more than one rank, in at least one call to
+      // run_algorithm (it is in addition also called with a single 'rank').
+      ParallelTestHelpers::runParallel(run_algorithm, 42, 42.0);
+    }
+  };
+
+Here ``MantidTestHelpers/ParallelAlgorithmCreation.h`` provides the algorithm factory method ``ParallelTestHelpers::create<WorkspaceType>``.
+``MantidTestHelpers/ParallelRunner.h`` provides ``ParallelTestHelpers::runParallel``, which uses ``ParallelRunner`` with a reasonable default choice for the number of ranks.
+
+Documentation
+-------------
+
+When adding MPI support for an algorithm, add it to the table at the end of this document.
+Potential limitations must be described in the comments.
+
+Supported Algorithms
+####################
+
+=============== =============== ========
+Algorithm       Supported modes Comments
+=============== =============== ========
+CreateWorkspace all
+Rebin           all             min and max bin boundaries must be given explicitly
+=============== =============== ========
+
+.. rubric:: Footnotes
+
+.. [#split-instrument] The complexity and overhead of splitting the instrument, in particular given the overhead ensuing from handling all cases exemplified above, led to the decision split only the neutron data based on spectra, but not detectors.
+
+.. [#spectrum-index] Some will argue that this should be ``GlobalWorkspaceIndex``.
+  However it is not an index of a workspace so the term ``GlobalSpectrumIndex`` has been chosen for clarity.
+  On the user interface side this will still be named 'workspace index', dropping the 'global' since the distinction between global and local indices is irrelevant for users.
+
+.. categories:: Development
```

Comment:
This document is a nice addition. We need to keep this up to date with future MPI work. Not sure what the best strategy for the reminder will be.